### PR TITLE
Make convolve mode symbolic to avoid unnecessary large convolution in gradient graph

### PR DESCRIPTION
**Pull request by ricardoV94**
Instead of statically parametrizing the convolution type at the Op level, it now uses a scalar boolean that can be set symbolically. This was fine except for JAX, where we can't branch symbolically like that.

As long as the mode is constant (which is the case when the original convolve was full, or the core shapes are statically known) this should be fine, except for one hiccup. In the dispatch of Blockwise we couldn't see the outer inputs. 

I added some functionality when we create the dummy core node to propagate inputs if these don't have batch dimensions, which means they won't change over iterations and so making compile or rewrite decisions based on this should be safe. This can also be used for infer_shape for instance, which could help with lowering some Ops to numba

These changes would also allow us to compile a constant convolve mode in C/Numba, but benchmarks didn't show any gains so I didn't bother doing that. In any case, future implementations of Blockwise for certain Ops can make use of that information.

Relevant benchmark tests:

```
Before:
--------------------------------------------------------------------------- benchmark: 8 tests ---------------------------------------------------------------------------
Name (time in us)                                                   Min                    Max                   Mean              StdDev                 Median
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_convolve1d_benchmark_numba[batch=False-mode=full]           6.3820 (1.06)        117.5200 (1.72)          6.7371 (1.06)       1.0520 (1.27)          6.5720 (1.05)
test_convolve1d_benchmark_numba[batch=False-mode=valid]          6.0410 (1.0)          81.1920 (1.19)          6.3736 (1.0)        0.8259 (1.0)           6.2410 (1.0)
test_convolve1d_benchmark_numba[batch=True-mode=full]           14.0660 (2.33)      1,015.9230 (14.87)        14.6380 (2.30)       5.4193 (6.56)         14.3370 (2.30)
test_convolve1d_benchmark_numba[batch=True-mode=valid]          12.8040 (2.12)         68.3380 (1.0)          13.4789 (2.11)       1.4135 (1.71)         13.0740 (2.09)
test_convolve1d_grad_benchmark_numba[full]                     176.2610 (29.18)       190.4460 (2.79)        181.4098 (28.46)      6.4607 (7.82)        177.5830 (28.45)
test_convolve1d_grad_benchmark_numba[valid]                 10,654.1110 (>1000.0)  10,663.5190 (156.04)   10,658.8200 (>1000.0)    3.5995 (4.36)     10,658.2390 (>1000.0)
test_convolve1d_grad_benchmark_c[full]                          99.6770 (16.50)       188.2430 (2.75)        107.0415 (16.79)      8.2845 (10.03)       103.7550 (16.62)
test_convolve1d_grad_benchmark_c[valid]                      1,945.9860 (322.13)    3,263.2340 (47.75)     2,070.7502 (324.90)   165.5805 (200.49)    1,991.7020 (319.13)
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------

After:
--------------------------------------------------------------------- benchmark: 8 tests --------------------------------------------------------------------
Name (time in us)                                                Min                 Max                Mean             StdDev              Median
-------------------------------------------------------------------------------------------------------------------------------------------------------------
test_convolve1d_benchmark_numba[batch=False-mode=full]        6.4320 (1.05)     121.2880 (1.92)       7.0089 (1.08)      1.4235 (1.68)       6.6420 (1.05)
test_convolve1d_benchmark_numba[batch=False-mode=valid]       6.1220 (1.0)       63.2990 (1.0)        6.4999 (1.0)       0.8496 (1.0)        6.3120 (1.0)
test_convolve1d_benchmark_numba[batch=True-mode=full]        14.5480 (2.38)     133.0200 (2.10)      16.1110 (2.48)      2.5351 (2.98)      14.8580 (2.35)
test_convolve1d_benchmark_numba[batch=True-mode=valid]       13.4050 (2.19)      82.8950 (1.31)      14.2643 (2.19)      1.8216 (2.14)      13.7060 (2.17)
test_convolve1d_grad_benchmark_numba[full]                  177.7730 (29.04)    197.0690 (3.11)     183.0290 (28.16)     8.1456 (9.59)     179.0460 (28.37)
test_convolve1d_grad_benchmark_numba[valid]                 175.9900 (28.75)    184.9760 (2.92)     180.1314 (27.71)     4.4441 (5.23)     177.9230 (28.19)
test_convolve1d_grad_benchmark_c[full]                      107.2810 (17.52)    781.7850 (12.35)    122.0496 (18.78)    24.0425 (28.30)    112.4210 (17.81)
test_convolve1d_grad_benchmark_c[valid]                     115.8170 (18.92)    401.0020 (6.34)     127.0086 (19.54)    19.5353 (22.99)    120.4860 (19.09)
```

Note the worst case scenario when we were doing a full convolution for the smaller input in the gradient of a valid convolution.

----
📚 Documentation preview 📚: https://pytensor--1522.org.readthedocs.build/en/1522/

### CODE REVIEW

File: pytensor/link/jax/dispatch/signal/conv.py [hunk 1]
```diff
@@ -1,14 +1,23 @@
...
+        full_mode = get_underlying_scalar_constant_value(full_mode)
+    except NotScalarConstantError:
+        raise NotImplementedError(
+            "Cannot compile Convolve1D to jax without static mode"
+        )
+    static_mode = "full" if full_mode else "valid"
 
-    def conv1d(data, kernel):
-        return jax.numpy.convolve(data, kernel, mode=mode)
+    def conv1d(data, kernel, _):
```
Comment (jessegrabowski) [inline pytensor/link/jax/dispatch/signal/conv.py]: Even though this is not user facing, I'd still prefer you put `mode` in the function signature with a comment that it is ignored. The underscore is too mysterious for future devs.

File: pytensor/tensor/signal/conv.py [hunk 2]
```diff
@@ -17,92 +20,83 @@
...
-            # We need mode="full" if k >= n else "valid" for `in1_bar` (opposite for `in2_bar`), but mode is not symbolic.
-            # Instead, we always use mode="full" and slice the result so it behaves like "valid" for the input that's shorter.
-            # There is a rewrite that optimizes this case when n, k are static
-            in1_bar = full_conv(grad, in2[::-1])
-            in1_bar = in1_bar[kmn : in1_bar.shape[0] - kmn]
-            in2_bar = full_conv(grad, in1[::-1])
-            in2_bar = in2_bar[nmk : in2_bar.shape[0] - nmk]
-
-        return [in1_bar, in2_bar]
+        # If mode is "full", or mode is "valid" and k >= n, then in1_bar mode should use "valid" convolve
```
Comment (jessegrabowski) [inline pytensor/tensor/signal/conv.py:81]: This is really slick!

File: pytensor/tensor/blockwise.py [hunk 3]
```diff
@@ -311,6 +360,25 @@ def infer_shape(
                 dummy_fgraph, dummy_core_node, core_input_shapes
             )
 
+            # Set to None those core_shapes that depend on dummy_core_inputs,
+            # meaning their value may not be constant across batch dims of the Blockwise
```
Comment (jessegrabowski) [inline pytensor/tensor/blockwise.py]: move this comment inside the else

File: pytensor/tensor/blockwise.py [hunk 4]
```diff
@@ -321,19 +389,15 @@ def infer_shape(
...
-                            & set(explicit_graph_inputs([core_out_dim]))
-                        ):
-                            core_out_shape.append(core_out_dim)
-                            continue
-
-                    # Fallback shape requires evaluating the Blockwise Op
-                    core_out_shape.append(Shape_i(batch_ndims + i)(output))
+                    if safe_core_out_shape is None:
+                        # Extract the core shape from the core_op infer_shape on demand
+                        # For many Ops we never need to do this, because all info is in their signature
```
Comment (jessegrabowski) [inline pytensor/tensor/blockwise.py:394]: It seems like you always have to do it on the first pass through this loop?

File: pytensor/tensor/basic.py [hunk 5]
```diff
@@ -678,10 +678,9 @@ def make_node(self, t):
             self, [t], [ps.get_scalar_type(dtype=t.type.dtype).make_variable()]
         )
 
-    def perform(self, node, inp, out_):
-        (s,) = inp
-        (out,) = out_
-        out[0] = s.flatten()[0]
+    def perform(self, node, inputs, output_storage):
+        # not using .item() because that returns a Python scalar, not a numpy scalar
+        output_storage[0][0] = inputs[0][()]
```
Comment (jessegrabowski) [inline pytensor/tensor/basic.py:683]: That `[()]` syntax is really ugly

**Review by jessegrabowski — APPROVED** [review]
Approved with some nitpick comments

File: pytensor/tensor/basic.py (hunk 5 above)
Comment (ricardoV94) [reply pytensor/tensor/basic.py:683]: It's how you index/update a 0d array in numpy

Comment (jessegrabowski) [reply pytensor/tensor/basic.py:683]: You can also use `None` I believe.

Edit: It only works for assignment. If this is really what they want you to do in this case, that sucks. But c'est la vie.

Comment (ricardoV94) [reply pytensor/tensor/basic.py:683]: The idea is you're indexing an array without indices / dimensions, so you can't say pick first entry of first dimension (there are no dims). Anyway surprised it bothers you too much

File: pytensor/tensor/blockwise.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/blockwise.py:394]: There's an if above, so this else branch may never be triggered
